Add specs for AcceleratorKeyPressed event args extension APIs (#3861)
MicrosoftEdge/WebView2Feedback: specs/ExtendedAcceleratorKeyPressed.md

# Examples
## C++

``` cpp
    EventRegistrationToken m_acceleratorKeyPressedToken = {};

    wil::com_ptr<ICoreWebView2Controller> m_controller;
    wil::com_ptr<ICoreWebView2> m_webView;
    wil::com_ptr<ICoreWebView2Settings3> m_settings3;

  {
        wil::com_ptr<ICoreWebView2Settings> settings;
        CHECK_FAILURE(m_webView->get_Settings(&settings));
        m_settings3 = settings.try_query<ICoreWebView2Settings3>();
        if (m_setting3) 
        {
            // Disable all browser accelerator keys
            CHECK_FAILURE(m_settings3->put_AreBrowserAcceleratorKeysEnabled(FALSE));
            // Register a handler for the AcceleratorKeyPressed event.
            CHECK_FAILURE(m_controller->add_AcceleratorKeyPressed(
                Callback<ICoreWebView2AcceleratorKeyPressedEventHandler>(
                    [this](
                        ICoreWebView2Controller* sender,
                        ICoreWebView2AcceleratorKeyPressedEventArgs* args) -> HRESULT
                    {
                        COREWEBVIEW2_KEY_EVENT_KIND kind;
                        CHECK_FAILURE(args->get_KeyEventKind(&kind));

>                         // We only care about key down events.

Reviewer: Is this accurate? Do we know that all browser accelerators are only ever handled in key down? Or are there any handled in key up?

Author: @bradp0721 This is part of the existing documentation we have for the current AcceleratorKeyPressed event, I wanted to keep this spec consistent with that, I would need to verify whether any of them are handled in key up.

Author: @bradp0721 ok, I confirmed they seem to be handled in key down only, even for the F7 and Ctrl+P and Key up is mapped to Release state to track for possible key combination.

                        if (kind == COREWEBVIEW2_KEY_EVENT_KIND_KEY_DOWN ||
                            kind == COREWEBVIEW2_KEY_EVENT_KIND_SYSTEM_KEY_DOWN)
                        {
                            UINT key;
                            CHECK_FAILURE(args->get_VirtualKey(&key));

                            wil::com_ptr<ICoreWebView2AcceleratorKeyPressedEventArgs2>
                                args2;

>                             CHECK_FAILURE(args->QueryInterface(IID_PPV_ARGS(&args2)));

Reviewer: This should probably gracefully handle the interface not being there, like we do for ICoreWebView2Settings3 above.

                            if (key == VK_F7)
                            {
                                // Allow the browser to process F7 key
                                CHECK_FAILURE(args2->put_AllowBrowserHandle(TRUE));
                            }
                        }
                        return S_OK;
                    })
                    .Get(),
                &m_acceleratorKeyPressedToken));

        }
    }
```

## C#
```c#
    CoreWebView2Settings _webViewSettings = webView.CoreWebView2.Settings;
    // All browser accelerator keys are disabled with this setting
    _webViewSettings.AreBrowserAcceleratorKeysEnabled = false;

    webView.CoreWebView2.AcceleratorKeyPressed += WebView2Controller_AcceleratorKeyPressed;

    void WebView2Controller_AcceleratorKeyPressed(object sender, CoreWebView2AcceleratorKeyPressedEventArgs e)
    {
        switch (e.KeyEventKind)
        {
            case CoreWebView2KeyEventKind.KeyDown:
            case CoreWebView2KeyEventKind.SystemKeyDown:
                {

>                     Keys keyData = (Keys)(e.VirtualKey);

Reviewer: Unused?

    }

```

# Remarks
By default, `AreBrowserAcceleratorKeyEnabled` is `TRUE` and default value for `AllowBrowserHandle` is `TRUE`. When developers set `AreBrowserAcceleratorKeyEnabled` to `FALSE` this will also change the default value for `AllowBrowserHandle` to `FALSE`. If developers want specific keys to be handled by the browser after changing the `AreBrowserAcceleratorKeyEnabled` settings to `FALSE`, they need to enable them by setting `event_args->put_ AllowBrowserHandle(TRUE)`. This option will give the event arg higher priority over the settings when the browser decide whether to handle the keys.  

# API Details
## C++
```
/// This is a continuation of the ICoreWebView2AcceleratorKeyPressedEventArgs interface.
[uuid(45238725-3774-4cfe-931f-7985a1b5866f), object, pointer_default(unique)]
interface ICoreWebView2AcceleratorKeyPressedEventArgs2 : ICoreWebView2AcceleratorKeyPressedEventArgs {
  /// These APIs allow devs to enable/disable the browser from handling specific accelerator keys.
  /// By default, `AreBrowserAcceleratorKeyEnabled` is `TRUE` and `AllowBrowserHandle` is `TRUE`.
  /// When devs set `AreBrowserAcceleratorKeyEnabled` to `FALSE` this will override and the

>   /// default value for `AllowBrowserHandle` will be `FALSE` and prevent the browser to handle

Reviewer 2: @bradp0721 Talked with Eve. This property would only apply to the same subset of accelerator keys that the existing CoreWebView2Settings.AreBrowserAcceleratorKeysEnabled property does (Ctrl+F yes, Ctrl+X no). So leaving as bool might be fine

>   /// default value for `AllowBrowserHandle` will be `FALSE` and prevent the browser to handle

Reviewer: Dave and I talked about this offline. This should only apply to the same browser accelerators covered by the AreBrowserAcceleratorKeysEnabled setting. This property essentially let's the app toggle the setting per browser accelerator.

>   /// default value for `AllowBrowserHandle` will be `FALSE` and prevent the browser to handle

Author: I did a quick test and confirmed that this only applies to the browser accelerators that AreBrowserAcceleratorKeysEnabled setting can enable/disable.

  /// accelerator keys. If users want specific keys to be handled by the browser after changing
  /// the settings to `FALSE`, they need to enable them by setting `AllowBrowserHandle` to `TRUE`.
  /// This option will give the event arg higher priority over the `AreBrowserAcceleratorKeyEnabled`

>   /// settings when we handle the keys.

Reviewer 2: Also, on the topic of the Handled property please also update the ref docs for it. It doesn't make it clear what exactly is getting marked handled or not (of the various things that might happen from accelerator key press)

  ///
  /// \snippet ScenarioAcceleratorKeyPressed.cpp
  /// Gets the `AllowBrowserHandle` property.

>   [propget] HRESULT AllowBrowserHandle([out, retval] BOOL* value);

Reviewer 2: The name: From the name I'm not sure what exactly I'm skipping if I disallow it: is it only skipping browser features associated with the accelerator (like how AreBrowserAcceleratorKeyEnabled handles Ctrl+F for find on page, but doesn't cover Ctrl+X for Cut), versus also skipping any feature associated with the accelerator (Ctrl+F _and_ Ctrl+X), versus also skipping the key press showing up in DOM keyboard events (Ctrl+F and Ctrl+X _and also_ the related DOM events aren't raised).

If it matches AreBrowserAcceleratorKeysEnabled we can name it similarly like AllowBrowserAcceleratorKeyHandling.
If its any accelerator key handling then we can name it like AllowDefaultAcceleratorKeyHandling.
And if its everything then we would probably name it Handled but that's already taken. So hopefully this isn't that behavior

Reviewer 2: @bradp0721 We talked about since this matches the subset of accelerator keys that the existing setting does it would be good to match that naming like `AllowBrowserAcceleratorKeyHandling`

Reviewer: I completely agree with renaming this to match the existing setting. The setting has Keys in the name as it covers all accelerator keys, so maybe we don't need Key since we're in an event for a specific key.
  AllowBrowserAccelerator
  AllowBrowserAcceleratorHandling
Or we make this the singular version of the setting.
  IsBrowserAcceleratorKeyEnabled
  IsBrowserAcceleratorEnabled

                    if (e.VirtualKey == Key.F7) 
                    {
                        e.AllowBrowserAcceleratorHandling = true;
                    }
                    break;
                }
            
        }
    }

```

# Remarks
By default, `AreBrowserAcceleratorKeysEnabled` is `TRUE` and default value for `AllowBrowserAcceleratorHandling`
is `TRUE`. When developers set `AreBrowserAcceleratorKeysEnabled` to `FALSE` this will also change the default 
value for `AllowBrowserAcceleratorHandling` to `FALSE`. If developers want specific keys to be handled by the 
browser after changing the `AreBrowserAcceleratorKeysEnabled` settings to `FALSE`, they need to enable them by 
setting `event_args->put_ AllowBrowserAcceleratorHandling(TRUE)`. This option will give the event arg higher 
priority over the settings when the browser decide whether to handle the keys.  

# API Details
## C++
```
/// This is a continuation of the ICoreWebView2AcceleratorKeyPressedEventArgs interface.
[uuid(45238725-3774-4cfe-931f-7985a1b5866f), object, pointer_default(unique)]
interface ICoreWebView2AcceleratorKeyPressedEventArgs2 : ICoreWebView2AcceleratorKeyPressedEventArgs {

>   /// Browser accelerator keys are the keys/key combinations that access features specific to 

Reviewer 2: You need the first sentence to be a summary of the property. Something like the following maybe

```suggestion
  /// This property allow you to enable or disable the browser from handling a specific browser 
  /// accelerator key such as Ctrl+P or F3.
  /// Browser accelerator keys are the keys/key combinations that access features specific to 
```

  /// a web browser, including but not limited to:
  ///  - Ctrl-F and F3 for Find on Page
  ///  - Ctrl-P for Print
  ///  - Ctrl-R and F5 for Reload
  ///  - Ctrl-Plus and Ctrl-Minus for zooming
  ///  - Ctrl-Shift-C and F12 for DevTools
  ///  - Special keys for browser functions, such as Back, Forward, and Search
  ///
  /// These APIs allow developers to enable/disable the browser from handling a specific browser 
  /// accelerator key. These APIs do not disable accelerator keys related to movement and text 
  /// editing, such as:
  ///  - Home, End, Page Up, and Page Down
  ///  - Ctrl-X, Ctrl-C, Ctrl-V
  ///  - Ctrl-A for Select All
  ///  - Ctrl-Z for Undo
  ///

>   /// The exsiting `AreBrowserAcceleratorKeysEnabled` API is a convenience setting for developers 

Reviewer 2: Throughout this ref doc comment please make sure you specify the class on which the property or event resides. Also for example applies to AcceleratorKeyPressed - should say CoreWebView2Controller.AcceleratorKeyPressed (is that the right class?)
 
```suggestion
  /// The exsiting `CoreWebView2Settings.AreBrowserAcceleratorKeysEnabled` API is a convenience setting for developers 

```

  /// to disable all the browser accelerator keys together. This setting sets the default value 
  /// for the `AllowBrowserAcceleratorHandling` property. 
  /// By default, `AreBrowserAcceleratorKeysEnabled` is `TRUE` and `AllowBrowserAcceleratorHandling` 
  /// is `TRUE`. 
  /// When developers change `AreBrowserAcceleratorKeysEnabled` setting to `FALSE`, this will 
  /// change default value for `AllowBrowserAcceleratorHandling` to `FALSE`. 
  /// If developers want specific keys to be handled by the browser after changing the 
  /// `AreBrowserAcceleratorKeysEnabled` setting to `FALSE`, they need to enable these keys by 
  /// setting `AllowBrowserAcceleratorHandling` to `TRUE`.
  /// This API will give the event arg higher priority over the `AreBrowserAcceleratorKeysEnabled` 
  /// setting when we handle the keys.
  ///
  /// For browser accelerator keys, when an accelerator key is pressed, the propagation and 
  /// processing order is: 

>   /// - An AcceleratorKeyPressed event is generated

Reviewer 2: ```suggestion
  /// - An AcceleratorKeyPressed event is raised
```

>   /// - Browser level handling

Reviewer 2: We might want to say WebView2 browser feature accelerator key handling to avoid confusion.
```suggestion
  /// - WebView2 browser feature accelerator key handling
```

  /// - Web Content Handling: If the key combination isn't reserved for browser actions, 
  /// the key event propagates to the web content, where JavaScript event listeners can 
  /// capture and respond to it.
  /// 

>   /// Currently, `ICoreWebView2AcceleratorKeyPressedEventArgs` has a `Handled` property, 

Reviewer 2: ```suggestion
  /// `ICoreWebView2AcceleratorKeyPressedEventArgs` has a `Handled` property, 

```

  /// that developers can use to marked a key as handled. When the key is marked as handled 
  /// anywhere along the path, the propagtion stops, and web content will not receive the key.
  /// With `AllowBrowserAcceleratorHandling` property, when developers mark 
  /// `AllowBrowserAcceleratorHandling` as `FALSE`, the browser will skip the browser level 

>   /// handling process, but the propagation continue, web content will receive the key 

Reviewer 2: ```suggestion
  /// handling process, but the event propagation continues, and web content will receive the key 
```

# Background
WebView2 has `ICoreWebView2AcceleratorKeyPressedEventArgs` for the 
`CoreWebView2Controller.AcceleratorKeyPressed` event. We have been asked to extend this API so

> that developers can indicate that theywould like an existing browser accelerator key to be 

Author: There's a main difference between this new property and Handled property that is explained in the API's comment section below. I will update this description to include that information to avoid confusion.

> that developers can indicate that theywould like an existing browser accelerator key to be 

Author: updated!

> that developers can indicate that theywould like an existing browser accelerator key to be 

Author: updated!

disabled as a browser accelerator.

In this document we describe the updated API. We'd appreciate your feedback.

# Description
We propose extending the `ICoreWebView2AcceleratorKeyPressedEventArgs` to allow developers
to enable/disable the browser from handling accelerator keys with `AllowBrowserAcceleratorHandling` 

> property. 

Reviewer 3: Confusing word order because it sounds like it's saying "This lets you control whether the browser handles accelerator keys that have the `AllowBrowserAcceleratorHandling` property. (But no effect on accelerator keys that don't have the `AllowBrowserAcceleratorHandling` property.)" But really it's

> We propose extending the `ICoreWebView2AcceleratorKeyPressedEventArgs` with a new `AllowBrowserAcceleratorHandling` property to allow developers to control whether the browser handles accelerator keys.

Author: updated!

# Examples
## C++

``` cpp
    EventRegistrationToken m_acceleratorKeyPressedToken = {};

    wil::com_ptr<ICoreWebView2Controller> m_controller;
    wil::com_ptr<ICoreWebView2> m_webView;
    wil::com_ptr<ICoreWebView2Settings3> m_settings3;

>   {

Reviewer 3: ```suggestion
   {
```
Believe it or not, I got confused by the mis-indentation.

Author: fixed!

>   {

Reviewer 3: This block of code probably should have a function declaration in front of it.
```cpp
    void MyWebView::InitializeAcceleratorHandling()
    {
        ...
    }
```

Author: updated!

            CHECK_FAILURE(m_settings3->put_AreBrowserAcceleratorKeysEnabled(FALSE));
            // Register a handler for the CoreWebView2Controller.AcceleratorKeyPressed event.
            CHECK_FAILURE(m_controller->add_AcceleratorKeyPressed(
                Callback<ICoreWebView2AcceleratorKeyPressedEventHandler>(
                    [this](
                        ICoreWebView2Controller* sender,
                        ICoreWebView2AcceleratorKeyPressedEventArgs* args) -> HRESULT
                    {
                        COREWEBVIEW2_KEY_EVENT_KIND kind;
                        CHECK_FAILURE(args->get_KeyEventKind(&kind));
                        // We only care about key down events.
                        if (kind == COREWEBVIEW2_KEY_EVENT_KIND_KEY_DOWN ||
                            kind == COREWEBVIEW2_KEY_EVENT_KIND_SYSTEM_KEY_DOWN)
                        {
                            UINT key;
                            CHECK_FAILURE(args->get_VirtualKey(&key));

                            wil::com_ptr<ICoreWebView2AcceleratorKeyPressedEventArgs2>
                                args2;

                            args->QueryInterface(IID_PPV_ARGS(&args2));

>                             if (args) 

Reviewer 3: ```suggestion
                            if (args2) 

```

Author: fixed!

                            {
                                if (key == VK_F7)
                                {
                                    // Allow the browser to process F7 key
                                    CHECK_FAILURE(args2->put_AllowBrowserAcceleratorHandling(TRUE));
                                }
                            }
                        }
                        return S_OK;
                    })
                    .Get(),
                &m_acceleratorKeyPressedToken));

        }
    }
```

## C#
```c#
    CoreWebView2Settings _webViewSettings = webView.CoreWebView2.Settings;
    // All browser accelerator keys are disabled with this setting. The default value for the 

>     // `AllowBrowserAcceleratorHandling`property is set to the current value of this setting. 

Author: fixed

    // AllowBrowserAcceleratorHandling to `false`
    _webViewSettings.AreBrowserAcceleratorKeysEnabled = false;

    webView.CoreWebView2.AcceleratorKeyPressed += WebView2Controller_AcceleratorKeyPressed;

    void WebView2Controller_AcceleratorKeyPressed(object sender, CoreWebView2AcceleratorKeyPressedEventArgs e)
    {
        switch (e.KeyEventKind)
        {
            case CoreWebView2KeyEventKind.KeyDown:
            case CoreWebView2KeyEventKind.SystemKeyDown:
                {
                    // Allow the browser to process F7 key
                    if (e.VirtualKey == Key.F7) 
                    {
                        e.AllowBrowserAcceleratorHandling = true;
                    }
                    break;
                }
            
        }
    }

```

# Remarks
By default, `CoreWebView2Settings.AreBrowserAcceleratorKeysEnabled` is `TRUE` and default value for
`AllowBrowserAcceleratorHandling` is `TRUE`. When developers set  
`CoreWebView2Settings.AreBrowserAcceleratorKeysEnabled` to `FALSE` this will also change the default 
value for `AllowBrowserAcceleratorHandling` to `FALSE`. If developers want specific keys to be handled 

> by the browser after changing the `CoreWebView2Settings.AreBrowserAcceleratorKeysEnabled` settings to 

Reviewer 3: The revised text should say simply

> The AcceleratorKeyPressed event is raised any time an accelerator key is pressed, regardless of whether accelerator keys are enabled or not.

(Note that we say that events are "raised", not "fired".)

> by the browser after changing the `CoreWebView2Settings.AreBrowserAcceleratorKeysEnabled` settings to 

Author: updated!

`FALSE`, they need to enable them by setting `event_args->put_ AllowBrowserAcceleratorHandling(TRUE)`. 
This option will give the event arg higher priority over the settings when the browser decide whether 
to handle the keys.  

# API Details
## C++
```
/// This is a continuation of the ICoreWebView2AcceleratorKeyPressedEventArgs interface.
[uuid(45238725-3774-4cfe-931f-7985a1b5866f), object, pointer_default(unique)]
interface ICoreWebView2AcceleratorKeyPressedEventArgs2 : ICoreWebView2AcceleratorKeyPressedEventArgs {
  /// This property allows developers to enable or disable the browser from handling a specific 
  /// browser accelerator key such as Ctrl+P or F3, etc.
  ///
  /// Browser accelerator keys are the keys/key combinations that access features specific to 

>   /// a web browser, including but not limited to:

Reviewer 4: All of this info in this long comment block here is great. I didn't really understand what this feature was about after reading the Background and Description sections. It was only when I read this comment block that I understood.

Would it be better to move some of this info out of the api comment and into the main section of the doc? 

I don't have strong opinions either way, but we should make sure that this info shows up clearly on the docs.microsoft.com page for this api.

Reviewer 3: The interaction between the AllowBrowserAcceleratorHandling and Handled properties is very important to understand and should be called out in real docs, not hidden in a comment.

Author: updated!

  ///  - Ctrl-F and F3 for Find on Page
  ///  - Ctrl-P for Print
  ///  - Ctrl-R and F5 for Reload
  ///  - Ctrl-Plus and Ctrl-Minus for zooming
  ///  - Ctrl-Shift-C and F12 for DevTools
  ///  - Special keys for browser functions, such as Back, Forward, and Search
  ///
  /// This property does not disable accelerator keys related to movement and text editing,
  /// such as:
  ///  - Home, End, Page Up, and Page Down
  ///  - Ctrl-X, Ctrl-C, Ctrl-V
  ///  - Ctrl-A for Select All
  ///  - Ctrl-Z for Undo
  ///

>   /// The exsiting `CoreWebView2Settings.AreBrowserAcceleratorKeysEnabled` API is a convenience setting 

Author: updated!

  /// for developers to disable all the browser accelerator keys together. This setting sets the default 
  /// value for the `AllowBrowserAcceleratorHandling` property. 
  /// By default, `CoreWebView2Settings.AreBrowserAcceleratorKeysEnabled` is `TRUE` and  
  /// `AllowBrowserAcceleratorHandling` is `TRUE`. 
  /// When developers change `CoreWebView2Settings.AreBrowserAcceleratorKeysEnabled` setting to `FALSE`,  
  /// this will change default value for `AllowBrowserAcceleratorHandling` to `FALSE`. 
  /// If developers want specific keys to be handled by the browser after changing the 
  /// `CoreWebView2Settings.AreBrowserAcceleratorKeysEnabled` setting to `FALSE`, they need to enable  
  /// these keys by setting `AllowBrowserAcceleratorHandling` to `TRUE`.
  /// This API will give the event arg higher priority over the  
  /// `CoreWebView2Settings.AreBrowserAcceleratorKeysEnabled` setting when we handle the keys.
  ///

>   /// For browser accelerator keys, when an accelerator key is pressed, the propagation and 

Author: updated!

  /// processing order is: 
  /// - A CoreWebView2Controller.AcceleratorKeyPressed event is raised
  /// - WebView2 browser feature accelerator key handling
  /// - Web Content Handling: If the key combination isn't reserved for browser actions, 
  /// the key event propagates to the web content, where JavaScript event listeners can 
  /// capture and respond to it.
  /// 
  /// `ICoreWebView2AcceleratorKeyPressedEventArgs` has a `Handled` property, that developers
  /// can use to marked a key as handled. When the key is marked as handled anywhere along

>   /// the path, the event propagtion stops, and web content will not receive the key.

Author: fixed!

  /// With `AllowBrowserAcceleratorHandling` property, when developers mark 
  /// `AllowBrowserAcceleratorHandling` as `FALSE`, the browser will skip the WebView2 
  /// browser feature accelerator key handling process, but the event propagation    
  /// continues, and web content will receive the key combination. 

  /// \snippet ScenarioAcceleratorKeyPressed.cpp
  /// Gets the `AllowBrowserAcceleratorHandling` property.
  [propget] HRESULT AllowBrowserAcceleratorHandling([out, retval] BOOL* value);

  /// Sets the `AllowBrowserAcceleratorHandling` property.
  [propput] HRESULT AllowBrowserAcceleratorHandling([in] BOOL value);
}
```

## C#
```c#
namespace Microsoft.Web.WebView2.Core
{
    runtimeclass CoreWebView2AcceleratorKeyPressedEventArgs
    {

>         Boolean AllowBrowserAcceleratorHandling { get; set; };

Reviewer 4: I find the name a little confusing. The name is very general ("Allow Browser Accelerator Handling"), but its meaning is very specific ("Allow the browser to handle this specific key press event"). 

Maybe something more specific:
>AllowBrowserToHandleThisEvent
>AllowBrowserToHandleThisKeyPress

Reviewer 3: I think "AllowBrowserToHandle..." is not quite accurate, because even if you set it to false, the key does go to the web page. It just gets delivered as a plain key and does not have special accelerator powers.

How about `TreatAsAccelerator`? Or if that's confusingly verb-like, `IsTreatedAsAccelerator`.

Reviewer 2: We want to say 'browser' accelerator to match the existing property name in order to distinguish between kinds of accelerators. Some accelerator keys will still be handled as accelerator keys, just not the ones that would be handled by browser specific features like the Browser's DevTools or print dialog accelerator keys.

How about 

* IsTreatedAsBrowserAccelerator

Author: I updated the naming to `IsBrowserAcceleratorKeyEnabled`, but if `IsTreatedAsBrowserAccelerator` is better/less confusing for this property, we can change it to `IsTreatedAsBrowserAccelerator`

Author: @kmahone @oldnewthing We changed the name of the API to `IsBrowserAcceleratorKeyEnabled`, does that naming better describe this API?

>         Boolean AllowBrowserAcceleratorHandling { get; set; };

Author: I updated the naming to IsBrowserAcceleratorKeyEnabled, but there's also a discussion above this about the potential naming, if IsTreatedAsBrowserAccelerator is better/less confusing for this property, we can change it to IsTreatedAsBrowserAccelerator
